### PR TITLE
feat(aws-iot): topic rules and actions

### LINKED ISSUE #5420: support lambda actions for IoT topic rules
Very similar to #555 

There does not appear to be an automatic way to add permissions (the "Function Policy") to the Lambda which is invoked from an IoT Topic Rule.

I obviously was not expecting the below to work since there is no explicit link between the IoT rule and the Lambda (since IoT does not appear to be supported by `lambda-event-sources`)

- The IoT rule is created with Lambda as its target
- The Lambda is created with __no trigger__ (nothing visible on the console)
- No Function Policy exists on the Lambda

Upon "editing" the IoT rule through the console, AWS helpfully says "we'll handle the Lambda permissions for you" - would be nice if this happened through the CDK

### Reproduction Steps

```ts
const motaAckLambda = new lambda.Function(this, 'MotaGwAck', {
    code: lambda.Code.fromAsset('apps/stacks/hw-mgmt/fota/dist/'),
    handler: 'fotaAck.handler',
    runtime: lambda.Runtime.NODEJS_10_X,
});
const lambdaIotAction: LambdaActionProperty = {
    functionArn: motaAckLambda.functionArn,
};
const iotFwdRule = new iot.CfnTopicRule(this, 'IotLambFwdRule', {
    topicRulePayload: {
        actions: [
            {
                lambda: lambdaIotAction,
            },
        ],
        ruleDisabled: false,
        sql: `SELECT soemthing FROM 'somewhere'`,
        awsIotSqlVersion: '2016-03-23',
    },
});
```

### Error Log

The IoT rule does NOT trigger the lambda and, in fact, I don't think the rule triggers at all (no error is logged in the verbose IoT logs in CloudWatch)

### Environment

  - **CLI Version      :1.18.0**
  - **Framework Version: 1.18.0**
  - **OS               : Mac**
  - **Language         : TypeScript**

#### Removing the policy 

- Whilst experimenting with this (to reproduce), if you need to remove the policy (you cannot do it through the console)

```
➜  ~ aws lambda get-policy --function-name functionArn
# Get the statement SID from here
➜  ~ aws lambda remove-permission --function-name functionArn --statement-id retrievedAbove
```

#### Current approach

My current approach (which I can confirm does work)

Everything as above and append:

```ts
motaAckLambda.addPermission('AllowIoTInvoke', {
    principal: new ServicePrincipal('iot.amazonaws.com'),
    sourceArn: iotFwdRule.attrArn,
});
```

It's unclear if this is the suggested fix (because IoT has notoriously never had a huge amount of documentation) but its working. Still, would be nice if IoT worked in the same way as almost every other Lambda event trigger

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by hollanddd**
Adds construct for IoT Topic Rule and introduces `@aws-iot/aws-iot-actions`. 

The supported Actions are Lambda, SNS, RepublishTopic & SQS.

This is a bit better and needs review when time permits:
`TopiRule.grantPublish` accepts `ITopicRule` and the `topic` string to publish to an MQTT topic. IoT topics are [maintained by the system designer](https://docs.aws.amazon.com/iot/latest/developerguide/topics.html) and do not contain a `CfnTopic` construct backing.
 
Closes #5420

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-iot/lib/topic-rule.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    return new Import(scope, id);
+  }
+
+  public readonly ruleName: string;
+  public readonly topicRuleArn: string;
+  private readonly actions = new Array<CfnTopicRule.ActionProperty>();
+
+  constructor(scope: Construct, id: string, props: TopicRuleProps) {
+    super(scope, id, {
+      physicalName: props.ruleName || 'TODO',
```
Comment (hollanddd) [inline packages/@aws-cdk/aws-iot/lib/topic-rule.ts:109]: How to ruleName when it's property is optional

File: packages/@aws-cdk/aws-iot-actions/lib/republish.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * Publishes to a IoT Topic
+ */
+export class Republish implements iot.ITopicRuleAction {
+  constructor(private readonly props: RepublishProps) {
+  }
+
+  public bind(rule: iot.ITopicRule): iot.TopicRuleActionConfig {
+    // Allow rule to publish to topic
+    const role = this.props.role || singletonTopicRuleRole(rule, [new iam.PolicyStatement({
```
Comment (hollanddd) [inline packages/@aws-cdk/aws-iot-actions/lib/republish.ts:43]: borrowed from [aws-event-target](https://github.com/aws/aws-cdk/blob/165a3d877b7ab23f29e42e1e74ee7c5cb35b7f24/packages/%40aws-cdk/aws-events-targets/lib/util.ts#L17) and have clear gaps in understanding or the role mechanism in this case.

**Comment by skinny85** [comment]
Thanks for the contribution @hollanddd! We have reviewed this pull request and marking it as p2, which means that we are unable to work on it immediately. We especially value community contributed code, so this is not a final decision.
We use +1s to help prioritize our work, and are happy to revaluate this pull request based on community feedback. You can reach out to the cdk.dev community on [Slack](https://join.slack.com/t/cdk-dev/shared_invite/zt-mso6p56d-qJp7SOTBvMaQuDrx7R2wHg) to solicit support for reprioritization.

**Comment by TheRealAmazonKendra** [comment]
This PR has been deemed to be abandoned, and will be automatically closed. Please create a new PR for these changes if you think this decision has been made in error.

**Pull request closed without merging**